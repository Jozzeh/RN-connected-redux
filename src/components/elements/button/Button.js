import React from 'react';
import {Text, TouchableHighlight} from 'react-native';

import {styles} from './Button.style';

/**
 * Use `Button` to perform an action in the system.
 */
const Button = ({testTextID, testID, title, onPress, enabled}) => {
  /* the testID is a prop because we want to test it in another component as well */
  return (
    <TouchableHighlight
      testID={testID ? testID : 'button-testid'}
      style={[styles.button]}
      underlayColor="darkcyan"
      onPress={() => {
        if (enabled) {
          onPress();
        }
      }}>
      <Text
        testID={testTextID ? testTextID : 'button-text-testid'}
        style={[styles.title]}>
        {title.toUpperCase()}
      </Text>
    </TouchableHighlight>
  );
};

export default Button;
