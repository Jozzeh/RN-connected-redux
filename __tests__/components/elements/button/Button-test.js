import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';

import Button from '../../../../src/components/elements/button/Button';

/* REDUX */
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk]; // add your middlewares like `redux-thunk`
const mockStore = configureStore(middlewares);

describe('Button element', () => {
  let store;
  let fakeFunction = jest.fn();

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = jest.fn();

    // THIS WILL CLEAR THE COUNTER OF HOW MANY TIMES IT HAS BEEN CALLED
    fakeFunction.mockClear();
  });

  /*
   * CHECK IF IT RENDERS
   * HAS TEXT CONTENT MATCHING THE TITLE PROP
   * FIRES EVENT ONPRESS
   * */
  it('renders & fire press event', () => {
    const testVariable = 'TEST';

    const {queryByTestId} = render(
      <Provider store={store}>
        <Button
          onPress={() => {
            fakeFunction();
          }}
          title={testVariable}
          testID={'button-testing'}
          enabled={true}
        />
      </Provider>,
    );
    const buttontext = queryByTestId('button-text-testid');
    expect(buttontext.props.children).toBe(testVariable.toUpperCase());

    const button = queryByTestId('button-testing');
    fireEvent.press(button);
    expect(fakeFunction).toHaveBeenCalledTimes(1);
    fireEvent.press(button);
    expect(fakeFunction).toHaveBeenCalledTimes(2);
  });

  it('renders alternatives & fire press event on disabled button', async () => {
    // RENDERS WITH testTextID
    render(
      <Provider store={store}>
        <Button testTextID="button-text-testid" title={'test'} enabled={true} />
      </Provider>,
    );

    // RENDERS WITH enabled FALSE
    const {queryByTestId} = render(
      <Provider store={store}>
        <Button
          onPress={() => {
            fakeFunction();
          }}
          title={'test'}
          enabled={false}
          testID={'button-testing'}
        />
      </Provider>,
    );
    const button = queryByTestId('button-testing');
    await fireEvent.press(button);
    expect(fakeFunction).toHaveBeenCalledTimes(0);
  });

  /* CREATE SNAPSHOTS ON ELEMENTS THAT DO NOT CHANGE OFTEN */
  it('matches snapshot', () => {
    const testVariable = 'TEST';

    const {toJSON} = render(
      <Provider store={store}>
        <Button title={testVariable} enabled={true} />
      </Provider>,
    );
    expect(toJSON).toMatchSnapshot();
  });
});
