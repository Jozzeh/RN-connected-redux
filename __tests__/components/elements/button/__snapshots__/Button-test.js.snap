// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button element matches snapshot 1`] = `[Function]`;
